README.md
# superfe-rn-inspector
[![NPM version][npm-image]][npm-url] [![Dependency status][david-dm-image]][david-dm-url] [![Dev Dependency status][david-dm-dev-image]][david-dm-dev-url]

Inspector tool for super-fe's react-native apps using [mocha](https://mochajs.org/).

```sh
npm install superfe-rn-inspector --save-dev
cd your_project
./node_modules/.bin/inspect .
```

This module can also be used programmatically from node:

```js
var inspect = require('superfe-rn-inspector');

inspect(process.cwd());
```

## changelog

Check out [CHANGELOG.md](CHANGELOG.md).

## license

BSD-2-Clause

[npm-url]: https://npmjs.org/package/superfe-rn-inspector
[npm-image]: http://img.shields.io/npm/v/superfe-rn-inspector.svg
[david-dm-url]:https://david-dm.org/super-fe/superfe-rn-inspector
[david-dm-image]:https://img.shields.io/david/super-fe/superfe-rn-inspector.svg
[david-dm-dev-url]:https://david-dm.org/super-fe/superfe-rn-inspector?type=dev
[david-dm-dev-image]:https://img.shields.io/david/dev/super-fe/superfe-rn-inspector.svg

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) 
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

## [2.2.2] - 2017-06-09
### Removed
- `private`

## [2.2.1] - 2017-03-22
### Changed
- fixed `check-node-version`

## [2.2.0] - 2017-03-22
### Removed
- check `react`&`react-native`

## [2.1.0] - 2017-03-04
### Changed
- target workspace default to current working directoru(cwd)
- perfer "contributors" than "auhtor"

## [2.0.0] - 2017-02-16
### Changed
- change fecs to eslint

## [1.2.5] - 2017-02-08
### Changed
- change node/npm version requirements

## [1.2.4] - 2017-02-06
### Changed
- "output" instead of "bundle" in `.gitignore`

## [1.2.3] - 2016-12-02
### Changed
- author could be string/array/object
- react-native could be in dev-dependencoes
### Added
- inspect react

## [1.2.2] - 2016-11-24
### Added
- inspect node & npm versions
- inspect .gitignore
- inspect .babelrc

## [1.2.0] - 2016-11-23
### Changed
- inspect local react-native version

## [1.1.1] - 2016-11-17
### Changed
- fixed wrong test names

## [1.1.0] - 2016-11-16
### Added
- split test cases into mutiple files
- validate version of react-native

## 1.0.0 - 2016-11-15
### Added
- initialize this project

[2.2.2]: https://github.com/super-fe/superfe-rn-inspector/compare/2.2.1...2.2.2
[2.2.1]: https://github.com/super-fe/superfe-rn-inspector/compare/2.2.0...2.2.1
[2.2.0]: https://github.com/super-fe/superfe-rn-inspector/compare/2.1.0...2.2.0
[2.1.0]: https://github.com/super-fe/superfe-rn-inspector/compare/2.0.0...2.1.0
[2.0.0]: https://github.com/super-fe/superfe-rn-inspector/compare/1.2.5...2.0.0
[1.2.5]: https://github.com/super-fe/superfe-rn-inspector/compare/1.2.4...1.2.5
[1.2.4]: https://github.com/super-fe/superfe-rn-inspector/compare/1.2.3...1.2.4
[1.2.3]: https://github.com/super-fe/superfe-rn-inspector/compare/1.2.2...1.2.3
[1.2.2]: https://github.com/super-fe/superfe-rn-inspector/compare/1.2.0...1.2.2
[1.2.0]: https://github.com/super-fe/superfe-rn-inspector/compare/1.1.1...1.2.0
[1.1.1]: https://github.com/super-fe/superfe-rn-inspector/compare/1.1.0...1.1.1
[1.1.0]: https://github.com/super-fe/superfe-rn-inspector/compare/1.0.0...1.1.0
[Unreleased]: https://github.com/super-fe/superfe-rn-inspector/compare/2.2.2...HEAD


tasks/eslint.js
/**
 * Copyright (C) 2016 baidu.com
 * eslint.js
 *
 * changelog
 * 2016-11-16[09:57:20]:revised
 * 2017-02-15[23:05:04]:rename to eslint
 *
 * @author yanni4night@gmail.com
 * @version 1.0.0
 * @since 0.1.0
 */
'use strict';
var Promise = require('bluebird');
var fs = Promise.promisifyAll(require('fs'));
var assert = require('assert');
var yaml = require('js-yaml');

describe('.eslintrc', function () {
    it('should exist', function () {
        assert.ok(fs.existsSync('.eslintrc'));
    });
    it('should be a JSON/YML format', function () {
        return fs.readFileAsync('.eslintrc', {
            encoding: 'utf-8'
        }).then(function (content) {
            var loaded = false;
            try {
                JSON.parse(content);
                loaded = loaded || true;
            } catch (e) {
                // empty
            }
            try {
                yaml.safeLoad(content);
                loaded = loaded || true;
            } catch (e) {
                // empty
            }
            assert.ok(loaded);
        });
    });
});

tasks/fecsrc.js
/**
 * Copyright (C) 2016 baidu.com
 * fecsrc.js
 *
 * changelog
 * 2016-11-16[09:57:20]:revised
 *
 * @author yanni4night@gmail.com
 * @version 0.1.0
 * @since 0.1.0
 */
'use strict';
var Promise = require('bluebird');
var fs = Promise.promisifyAll(require('fs'));
var assert = require('assert');
var yaml = require('js-yaml');

describe('.fecsrc', function () {
    it('should exist', function () {
        assert.ok(fs.existsSync('.fecsrc'));
    });
    it('should be a JSON/YML format', function () {
        return fs.readFileAsync('.fecsrc', {
            encoding: 'utf-8'
        }).then(function (content) {
            var loaded = false;
            try {
                JSON.parse(content);
                loaded = loaded || true;
            } catch (e) {
                // empty
            }
            try {
                yaml.safeLoad(content);
                loaded = loaded || true;
            } catch (e) {
                // empty
            }
            assert.ok(loaded);
        });
    });
});

tasks/babelrc.js
/**
 * Copyright (C) 2016 yanni4night.com
 * babelrc.js
 *
 * changelog
 * 2016-11-24[09:38:49]:revised
 *
 * @author yanni4night@gmail.com
 * @version 0.1.0
 * @since 0.1.0
 */
'use strict';
var Promise = require('bluebird');
var fs = Promise.promisifyAll(require('fs'));
const assert = require('assert');

describe('.babelrc', function () {
    it('should exist', function () {
        assert.ok(fs.existsSync('.babelrc'));
    });
    var babelrcContent;
    it('should be a JSON format', function () {
        return fs.readFileAsync('.babelrc', {
            encoding: 'utf-8'
        }).then(function (content) {
            babelrcContent = JSON.parse(content);
        });
    });
    it('should include a "presets" preset', function () {
        assert.ok(babelrcContent.presets, '"presets" exist');
        assert.ok(Array.isArray(babelrcContent.presets), 'is array format');
        assert.ok(babelrcContent.presets.indexOf('react-native') > -1, 'includes "react-native"');
    });
});

tasks/package.js
/**
 * Copyright (C) 2016 baidu.com
 * package.js
 *
 * changelog
 * 2016-11-15[17:03:19]:revised
 * 2017-03-22[17:10:32]:remove react&react-native
 *
 * @author yanni4night@gmail.com
 * @version 0.2.0
 * @since 0.1.0
 */
'use strict';
var Promise = require('bluebird');
var fs = Promise.promisifyAll(require('fs'));
var assert = require('assert');
var semver = require('semver');
var resolve = require('resolve');
var compatibleRnVersions = require('superfe-compatible-rn-versions');
var _ = require('lodash');

describe('package.json', function () {
    it('should exist', function () {
        assert.ok(fs.existsSync('package.json'));
    });
    var pkg;
    it('should be a JSON format', function () {
        return fs.readFileAsync('package.json', {
            encoding: 'utf-8'
        }).then(function (content) {
            return (pkg = JSON.parse(content));
        });
    });
    it('should have a "name" in string', function () {
        assert.ok(_.isString(pkg.name));
        assert.ok(_.trim(pkg.name).length > 0);
    });
    it('should have a valid version', function () {
        assert.ok(semver.valid(pkg.version));
    });
    // it('should have a "private=true"', function () {
    //     assert.deepEqual(pkg.private, true);
    // });
    it('should have a "scripts" in object', function () {
        assert.ok(_.isPlainObject(pkg.scripts));
    });
    it('should have a "scripts.test" in string', function () {
        assert.ok(_.isString(pkg.scripts.test));
        assert.ok(_.trim(pkg.scripts.test).length > 0);
    });
    it('should have a "pre-commit" in string', function () {
        assert.ok(_.isString(pkg['pre-commit']));
        assert.ok(_.trim(pkg['pre-commit']).length > 0);
    });
    it('value of "pre-commit" should be in "scripts"', function () {
        assert.ok(pkg['pre-commit'] in pkg.scripts);
    });
    it('should have a "author" or "contributors"', function () {
        var checkAuthorItem = function (author) {
            assert.ok(author);
            assert.ok(_.isString(author.name) && author.name, '"name" should be non-empty string');
            assert.ok(_.isString(author.email) && author.email,
                '"email" should be non-empty string');
        };

        assert.ok(('author' in pkg) || ('contributors' in pkg), 'none of them exist');

        if ('author' in pkg) {
            assert.ok(!_.isArray(pkg.author), 'You may change "author" to "contributors"');

            switch (true) {
            case _.isPlainObject(pkg.author):
                checkAuthorItem(pkg.author);
                break;
            case _.isString(pkg.author):
                assert.ok(pkg.author, '"author" should be non-empty');
                break;
            default:
                assert.ok(false, '"author" should be a string or a plain object');
            }
        }
        if ('contributors' in pkg) {
            assert.ok(_.isArray(pkg.contributors), '"contributors" should be an array');
            pkg.contributors.forEach(checkAuthorItem);
        }

    });
    it('should have a "dependencies" in object', function () {
        assert.ok(_.isPlainObject(pkg.dependencies));
    });
    it('local "react-native" should match ' + compatibleRnVersions.join('/'), function () {
        let pkg = resolve.sync('react-native/package.json', {
            basedir: process.cwd()
        });
        if (pkg) {
            pkg = require(pkg);
            try {
                assert.ok(compatibleRnVersions.indexOf(pkg.version) > -1);
            } catch (e) {
                this.skip();
            }
        } else {
            this.skip();
        }

    });
    it('should have a "devDependencies" in object', function () {
        assert.ok(_.isPlainObject(pkg.devDependencies));
    });
    /*it('should depend on "react-native"', function () {
        assert.ok('react-native' in pkg.dependencies || 'react-native' in pkg.devDependencies);
    });
    it('should depend on "react"', function () {
        assert.ok('react' in pkg.dependencies || 'react' in pkg.devDependencies);
    });*/
});